### PR TITLE
Rework connection classes with proper async support

### LINKED ISSUE #210: Synchronous operations inside async functions
> [ba1dr](https://github.com/ba1dr) **Authored by [ba1dr](https://github.com/ba1dr)**
_<time datetime="2022-02-13T17:59:22Z" title="Sunday, February 13th 2022, 11:59:22 am -06:00">Feb 13, 2022</time>_

---

mcstatus==8.0.0

I have a hanging process while I query servers in async mode. This happens not always - 1-2 times a day.
I managed to reproduce this only after 3 hours in infinite loop of polling 100+ servers.
See stacktrace below.

```
...
  File "/opt/myfolder/some_file.py", line 145, in async_query
    return await mcserver.async_query(tries=2)
  File "/usr/local/lib/python3.9/dist-packages/mcstatus/server.py", line 193, in async_query
    await querier.handshake()
  File "/usr/local/lib/python3.9/dist-packages/mcstatus/querier.py", line 70, in handshake
    self.challenge = int(packet.read_ascii())
  File "/usr/local/lib/python3.9/dist-packages/mcstatus/protocol/connection.py", line 80, in read_ascii
    result.extend(self.read(1))
KeyboardInterrupt
```

As I can see there are lots of synchronous operations in functions defined as `async`.

```
...
class AsyncServerQuerier(ServerQuerier):
    async def _read_packet(self):
        packet = Connection()
...

    async def handshake(self):
        await self.connection.write(self._create_handshake_packet())

        packet = await self._read_packet()
        self.challenge = int(packet.read_ascii())

...
```

here `read_ascii()` is a sync operation.

Next: `QueryResponse.from_connection()` is sync too.

And so on..

### PR DESCRIPTION
**Pull request by CoolCat467**
Rebase of #225.
Closes #210 

This PR introduces a complete rewrite of the protocol classes and fixes a bug, in which asynchronous connections were sometimes using synchronous functions when asynchronous ones were expected. 

It adds doc strings to all of the methods, introduces some memory optimizations with the use of `__slots__` in the connection classes, and separates `Connection` into several abstract classes, both synchronous and asynchronous, which then together compose the sync and asynchronous Connection classes. These abstract classes prevent a lot of previously duplicate code.

TODO:
- [x] Comply with linters, formatters and type checkers (black, isort, flake8, pyright)
- [x] Fix unit tests for the protocol classes to work with this rewrite
- [x] ~~Adjust the rest of the code base to work with this rewritten code~~ No changes should be required, but improvements could be made

### CODE REVIEW

**Comment by ItsDrike** [comment]
Hi @CoolCat467, thanks for the rebase (btw, you don't need to make a new PR when rebasing, we do allow force pushing). I've edited your PR description to include some more detailed things about what this is doing (mostly based on the previous PR description), but most importantly, I've added some TODO check boxes there which needs to be done before this can be reviewed and then merged.

I've quickly glanced over the changes this introduces and it looks pretty interesting, but until it meets these TODOs and is compliant with our code quality standards, it can't be merged. Because of that, I've marked your PR as draft to avoid cluterring other PRs which are actually ready to be merged already. Once you feel like the PR is ready to be reviewed and you fulfilled these points and are passing the automated workflow checks, you can mark it as Ready for review (this button):

image

If you need more info on how to fulfill some of these, or how to pass the automated workflows, feel free to ask on discord, or here as a comment (though discord will probably be better for longer discussions).

**Comment by CoolCat467** [comment]
Ok, thank you for the information.

**Comment by ItsDrike** [comment]
@CoolCat467, do you plan on going through with this and implementing it fully or should someone take over this PR?

**Comment by CoolCat467** [comment]
I've never written unit tests before, so that would probably be for the best.

**Comment by ItsDrike** [comment]
> I've never written unit tests before, so that would probably be for the best.

Alright, no problem, I don't mind finishing this up, although it may take me a while since I've been a bit busy lately, so if someone else would want to, feel free to take over, otherwise I'll implement this hopefully within in a few weeks.

File: mcstatus/protocol/connection.py
Comment (kevinkjt2000) [inline mcstatus/protocol/connection.py:3]:
```suggestion
```
No need for the `#!` since this is not an executable script.
Python3 files are assumed to be encoded in utf-8. This magic comment is a remnant from Python2 where strings were encoded differently by default.

File: mcstatus/protocol/connection.py
Comment (kevinkjt2000) [inline mcstatus/protocol/connection.py:20]:
```suggestion
if TYPE_CHECKING:
    from typing_extensions import SupportsIndex  # Python 3.7 doesn't support this yet.

    BytesConvertable = Union[SupportsIndex, Iterable[SupportsIndex]]
```
The conditional import is there so this does not cause an ImportError at runtime. mcstatus does not list `typing-extensions` as a runtime dependency.

File: mcstatus/protocol/connection.py
Comment (kevinkjt2000) [inline mcstatus/protocol/connection.py:13]:
```suggestion
from typing import Iterable, Optional, TYPE_CHECKING, Union
```

**Review by kevinkjt2000 — CHANGES_REQUESTED** [review]
There are a handful of type checking errors when running `tox -e lint`. See CONTRIBUTING.md for the commands to set up your terminal. Once those type checks are fixed, I'd start running `tox -e py310` to execute tests for a specific version. When I get a single version working, I like to use both the oldest and newest `tox -e py37,py310`. To run the whole suite of format check, linting, and tests: `tox`.

If you ping me on the Discord server, I could schedule some time to teach you some unit test stuff if you're interested. Unit tests are basically a collection of a bunch of miniature Python scripts that have assertions sprinkled into them.

I'll wait until things are passing again to do a more thorough review of the new connection classes, but I'm on board with this idea. Thanks for taking the time on this PR! My apologies for taking a while to get around to this.

File: mcstatus/protocol/connection.py
Comment (CoolCat467) [reply mcstatus/protocol/connection.py:17]: That makes sense, but I'm confused how BytesConvertable is able to be used even though it shouldn't exist if TYPE_CHECKING is false

Comment (ItsDrike) [reply mcstatus/protocol/connection.py:17]: We're using `from futures import __annotations__` which automatically treats every annotation as a string which means it won't cause a runtime error

**Comment by CoolCat467** [comment]
Ok so I think I've gotten almost everything to a good state, but I've got two interesting issues on the unit tests.

First, one of the tests sets a `Connection` object's `remaining` to a `Mock` object, but it fails and says it's not allowed to do that and that it's read-only. I don't understand why it can't, it's not a property or anything, so it should be allowed to overwrite it, but no.

Second, a different test is wanting a TCP connection to raise an asyncio TimeoutError and I'm not sure how to fix that.

Other than that, as far as I understand it, everything _should_ work.

**Review by ItsDrike — CHANGES_REQUESTED** [review]
Just a quick review purely addressing the formatting/typing issues, I'll make another review on the actual logic after these changes are addressed. Thanks for the contribution though, the connection classes really needed an update for a while now!

File: mcstatus/protocol/connection.py
Comment (CoolCat467) [reply mcstatus/protocol/connection.py:560]: None would mean needing to ignore a type assignment though.

Comment (ItsDrike) [reply mcstatus/protocol/connection.py:560]: Yeah, but that's not a huge issue considering we already type-ignore the reader and writer variables, there's just no point in assigning some value here if it's always expected to be overridden later.

File: mcstatus/server.py
Comment (CoolCat467) [reply mcstatus/server.py]: Changed as of 68452e4

File: mcstatus/protocol/connection.py
Comment (ItsDrike) [reply mcstatus/protocol/connection.py:25]: You seem to have accidentally reverted this change in 68452e4, also this suggestion applies to all of the added docstrings, not just this one

File: mcstatus/protocol/connection.py
Comment (CoolCat467) [reply mcstatus/protocol/connection.py]: Changed to 0 in c92282f, is that sufficient?

Comment (ItsDrike) [reply mcstatus/protocol/connection.py]: I'd prefer `None`, since specifying a number here is misleading, suggesting that the timeout is hard-coded to certain value, whereas just setting it to None and type-ignoring with that comment above explaining why these are ignored is a bit better way to handle this.

Comment (CoolCat467) [reply mcstatus/protocol/connection.py]: Changed in c09d857

File: mcstatus/protocol/connection.py
Comment (CoolCat467) [reply mcstatus/protocol/connection.py]: Changed in c09d857

File: mcstatus/protocol/connection.py
Comment (PerchunPak) [inline mcstatus/protocol/connection.py:119]: There is a couple of methods with duplicated logic. Maybe you should do it in one method with dict or possibly enum?
```py
from enum import Enum

class ValueType(Enum):
    FULL_NAME = "Q"
    ...

def write_value_with_type(self, value: int, type: ValueType) -> None:
    self.write(self._pack(type.value, value))

write_value_with_type(123, ValueType.FULL_NAME)
```
https://docs.python.org/3/library/enum.html

File: mcstatus/protocol/connection.py
Comment (PerchunPak) [inline mcstatus/protocol/connection.py:55]:
```suggestion
        :param value: Max 2 ** 31 - 1, min -(2 ** 31).
```

Applies to all `max` and `min` mentions.

File: mcstatus/protocol/connection.py
Comment (PerchunPak) [inline mcstatus/protocol/connection.py:56]: We using RST format for docstrings, so you should use `:raises ValueError: ...`.
https://peps.python.org/pep-0287/#docstring-significant-features

```suggestion
        :raises ValueError: If varint is out of range.
```

Apllies to all non-RST docstring formatting.

File: mcstatus/protocol/connection.py
Comment (PerchunPak) [inline mcstatus/protocol/connection.py:643]: Are we using pylint?

**Review by PerchunPak — CHANGES_REQUESTED** [review]

File: mcstatus/protocol/connection.py
Comment (PerchunPak) [reply mcstatus/protocol/connection.py]: There's also a few more places, like this:

- [lines 246-248](https://github.com/py-mine/mcstatus/pull/273/files#diff-841cbf8f4a8df4e0442dbd7d12482953acca60f80e4022fda10fbdbf50ad1956R246-R248)
- [lines 258-260](https://github.com/py-mine/mcstatus/pull/273/files#diff-841cbf8f4a8df4e0442dbd7d12482953acca60f80e4022fda10fbdbf50ad1956R258-R260)
- [lines 345-347](https://github.com/py-mine/mcstatus/pull/273/files#diff-841cbf8f4a8df4e0442dbd7d12482953acca60f80e4022fda10fbdbf50ad1956R345-R347)

File: mcstatus/protocol/connection.py
Comment (PerchunPak) [reply mcstatus/protocol/connection.py]: Found some unchanged places:

- [line 509](https://github.com/py-mine/mcstatus/pull/273/files#diff-841cbf8f4a8df4e0442dbd7d12482953acca60f80e4022fda10fbdbf50ad1956R509)
- [line 69](https://github.com/py-mine/mcstatus/pull/273/files#diff-ae432ff1129444c75c1d54e0b41718eba44c2c678621698de2c48e64a73180dbL69-R69)

File: mcstatus/protocol/connection.py
Comment (ItsDrike) [reply mcstatus/protocol/connection.py:119]: Cool idea, but the original connection implementation was already using these methods, and this PR's goal is mainly to introduce async support, not necessarily to fix bad design patters in that code.

This change would be compatibility breaking and will require changes in the code-base and depracation periods on these old methods (although I'm not sure how many users of the library are actually using the connection class directly, I don't expect that number to be very big). I'll leave this up to the author to decide whether they're willing to implement this, if not, it can be done later in another PR.

Comment (PerchunPak) [reply mcstatus/protocol/connection.py:119]: `mcstatus` really give so low level API? I thought it's just a backend.

I was pretty sure all public API is `server.py` file and answer objects, because only it somehow documented in README.

Comment (ItsDrike) [reply mcstatus/protocol/connection.py:119]: Yeah you're right, deprecation wouldn't actually be necessary here. Still though, leaving this up to the author on whether they want to implement it here or not.

File: mcstatus/protocol/connection.py
Comment (CoolCat467) [reply mcstatus/protocol/connection.py:44]: There is never a "need" per say for a repr, but it's good practice to do it anyhow.

File: mcstatus/protocol/connection.py
Comment (CoolCat467) [reply mcstatus/protocol/connection.py:643]: No, but I made an IDLE extension for pylint that I've been using while writing this.

File: mcstatus/protocol/connection.py
Comment (CoolCat467) [reply mcstatus/protocol/connection.py:119]: That is a good idea, to avoid duplicated code, but I agree that it would be compatibility breaking. I think that would be better suited for a new PR instead of this one. This one changes things enough as is lol.

File: mcstatus/protocol/connection.py
Comment (CoolCat467) [reply mcstatus/protocol/connection.py]: Removed as of a09673e

File: mcstatus/protocol/connection.py
Comment (CoolCat467) [reply mcstatus/protocol/connection.py]: Changed in a09673e

File: mcstatus/protocol/connection.py
Comment (CoolCat467) [reply mcstatus/protocol/connection.py]: Changed in a09673e

File: mcstatus/protocol/connection.py
Comment (CoolCat467) [reply mcstatus/protocol/connection.py:44]: And why use dataclasses when this will suffice? Plus I feel like there are problems with __slots__ in dataclasses in older versions or something, and if this is networking code faster attribute access is important.

Comment (ItsDrike) [reply mcstatus/protocol/connection.py:44]: Yeah, slotted dataclasses are 3.10+, though you can jusr define `__slots__` explicitly, but this caused problems with inheritance iirc. I would agree that in this case dataclasses are an overkill.

Comment (ItsDrike) [reply mcstatus/protocol/connection.py:44]: This is now tracked in #355, marking this discussion as resolved

File: mcstatus/protocol/connection.py
Comment (ItsDrike) [reply mcstatus/protocol/connection.py:119]: This is now tracked in #355, marking this discussion as resolved.

File: mcstatus/server.py
Comment (ItsDrike) [inline mcstatus/server.py:41]: Annotations on `__init__`'s return aren't required as the type-checker can infer this. Also, this PR has no reason to make any changes in the server.py file, not sure why this is here.

```suggestion
    def __init__(self, host: str, port: Optional[int] = None, timeout: float = 3):
```

File: mcstatus/protocol/connection.py
Comment (ItsDrike) [inline mcstatus/protocol/connection.py:544]: Our maximum line length is 127, so this can easily fit into a single docstring line.
```suggestion
        """Set self.addr to address, set self.socket to new socket, AF_INET if IPv4, AF_INET6 otherwise."""
```

File: mcstatus/protocol/connection.py
Comment (ItsDrike) [inline mcstatus/protocol/connection.py:123]: Our maximum line length is 127, so this can easily fit into a single docstring line.

```suggestion
        """Flush buffer, then write a varint of the length of the buffer's data, then write buffer data."""
```

File: mcstatus/protocol/connection.py
Comment (ItsDrike) [inline mcstatus/protocol/connection.py:509]:
```suggestion
    """TCP Connection to address. Timeout defaults to 3 seconds."""
```

File: mcstatus/protocol/connection.py
Comment (ItsDrike) [inline mcstatus/protocol/connection.py:407]:
```suggestion
        return f"<{self.__class__.__name__} Object>"
```

File: mcstatus/protocol/connection.py
Comment (ItsDrike) [inline mcstatus/protocol/connection.py:141]:
```suggestion
        return f"<{self.__class__.__name__} Object>"
```

File: mcstatus/protocol/connection.py
Comment (ItsDrike) [inline mcstatus/protocol/connection.py:325]:
```suggestion
        return f"<{self.__class__.__name__} Object>"
```

File: mcstatus/protocol/connection.py
Comment (ItsDrike) [inline mcstatus/protocol/connection.py:629]:
```suggestion
        self.timeout: float = None  # type: ignore[assignment]
```

File: mcstatus/tests/test_server.py
Comment (ItsDrike) [inline mcstatus/tests/test_server.py:170]: It's a bit weird to directly access the mocked feature and set the side-effect rather than simply using `as` to obtain that mock.

```suggestion
        with patch("mcstatus.protocol.connection.Connection.remaining") as mock_remaining:
            mock_remaining.side_effect = [15, 208]
```

**Review by ItsDrike — CHANGES_REQUESTED** [review]
Great job! The code does still need a lot of minor formatting changes, but logic-wise I don't see any significant issues to point out (other than some that were already there before this PR, which should be fixed elsewhere anyway).

File: mcstatus/protocol/connection.py
Comment (CoolCat467) [reply mcstatus/protocol/connection.py:544]: Gotcha

File: mcstatus/server.py
Comment (CoolCat467) [reply mcstatus/server.py:41]: Using mypy to check things while working on this, and I've set up my config to raise warnings whenever functions don't have type annotations. I agree there is no reason to do this.

File: mcstatus/tests/test_server.py
Comment (CoolCat467) [reply mcstatus/tests/test_server.py:170]: I was thinking while working on this that it was a bit weird, and again I'm not super familiar with writing tests, so thanks for suggestion.

File: mcstatus/server.py
Comment (ItsDrike) [reply mcstatus/server.py]: Yeah, we're using `flake8-annotations` to enforce type-hints, and `pyright` as the type-checker for this project, `mypy` isn't entirely compatible with pyright and there are some newer typing features it doesn't yet support which are used in the code-base, such as the `Self` type. I'd recommed switching to this setup to keep things compatible with what's getting run by our automated tests.

**Review by ItsDrike — APPROVED** [review]

**Review by kevinkjt2000 — APPROVED** [review]

**Pull request merged**
